### ISSUE TITLE
Sporadic 504 timeouts since 1.22 upgrade

### ISSUE THREAD

**Issue by Stono** [istio/istio#51660]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Hello!
Back with another issue i'm struggling to diagnose since 1.22 upgrade.

Really infrequently, we're bursts of 504s.  A single pod, for a period of time (a few seconds), will fail to talk to **all** its upstreams.

The traces are _extra_ weird.  What we always see is a span that times out with a `504UT`, the timeout is the timeout configured on the destination services `VirtualService` (as you'd expect).    However the request does eventually get handled by that destination workload (and the span for its handling the request is rapid, as you'd expect.

Here's an example:
<img width="1728" alt="Screenshot 2024-06-21 at 15 17 47" src="https://github.com/istio/istio/assets/1486729/fe2dcc28-1fde-4fe0-90a4-f048d402d3d7">

Here's another example:
<img width="1728" alt="Screenshot 2024-06-21 at 15 19 05" src="https://github.com/istio/istio/assets/1486729/8055bfc4-ee8b-41b3-a3eb-f85775ccd8fa">

Countless examples of where for that brief moment, a singular source pod gets this behaviour with N destination workloads.

The problem is that it's so sporadic, like a couple of times per day, for a handful of requests out of many millions, on completely non-deterministic workloads.  As a result I have no chance at all of capturing it with a packet trace.

I appreciate there's just not enough information here to actually debug.  I've just spent 3 days looking into this and felt I should share it in case others had observed the same bahviour too.

I have ruled out everything possible that _isn't_ istio.  Today I've completed rolled back 1.22 to 1.21 everywhere, and the issues appear to have stopped.  However I won't know conclusively for another few days.

### Version

```Text
1.22.1
```

### Additional Information

Obviously this is an on going investigation for me so if i find anything else to help, i'll add to this thread.

**Comment by Stono** [comment]
Just to update here, over 48hrs since we reverted and we've not had any reoccurrence of this issue.

**Comment by howardjohn** [comment]
Looking at the spans, it looks like 90s of ingress active, then some delay, then the rest of the apps (quick). Huge period of time where nothing is active.

Is there some explanation for that behavior? or is that part of the issue

**Comment by Stono** [comment]
Yes that _is_ the issue. 
What we see, in the example above, would be:

- nginx -> app 👍 
- app logs inbound request 👍 
- app logs outbound request 👍 
- app hits its internal client timeout of say, 10s
- some time later, the request is sent and processed by the downstream system.

The best way i can describe is that there's something blocking in the source system sidecar, preventing that request from getting sent.

**Comment by Stono** [comment]
And this will happen for all requests to _one or more_ destination hosts from _a single pod_ for say 5-10s, then it'll all recover.

And to be clear it'll just be one pod, so in the above example lets say - there's 6 source and 6 destination, just one of the sources will stop being able to communicate outbound to N of its downstream systems.

**Comment by Stono** [comment]
I honestly don't know where to go from here, I kinda feel stuck, I've tried everything I could think of to reproduce it and haven't been able to do so.  It's so infrequent and only seems to occur in higher load environments.  Subsequently I can't possibly run a wireshark or similar for long enough to catch it.  

Equally I fully appreciate how here there isn't remotely enough information for Istio folks to be able to help me debug it either.  So I'm kinda 🤷

**Comment by Stono** [comment]
So a quick update here, with the release of `1.22.2`, I rolled forward again - but this time I set `ISTIO_DELTA_XDS: false` due to https://github.com/istio/istio/issues/51612.

So far I haven't seen a reoccurrence of this issue which means one of:

- It's some interaction with Delta XDS
- It was fixed in 1.22.2
- It was nothing to do with istio and some other network level problem was manifesting at exactly the same time and stopped when i rolled back (unlikely - but can't prove it!)

**Comment by Stono** [comment]
Spoke too soon, just had a batch on 1.22.2, single pod, single node, other things on that node not affected therefore you can infer its a pod issue.

![Screenshot 2024-07-02 at 19 08 53](https://github.com/istio/istio/assets/1486729/334f1711-32f4-4111-82f7-a6944269e839)

I do in parallel have an open ticket with GCP because we're seeing other _strange_ things today that _feel_ networky (i know....), but this one to me feels really isolated to the pod.

**Comment by jpuskar** [comment]
We had something similar happen. We just upgraded to 1.22.3 from a much older version.

One pod and only one pod of a deployment was timing out to downstream.
Pod logs seem fine; pod was receiving traffic.
We scaled up the deployment and then everything started working, including the original broken pod that was broken for > 30mins while we were troubleshooting.

We've looked through all our logs and can't make heads or tails of it.

So something about the act of scaling up the deployment fixed the outbound traffic of our broken pod. My hypothesis is that istiod updated something in the sidecar of the broken pod, when new pods were added.

**Comment by jpuskar** [comment]
@Stono, did you ever figure this one out? I can't figure out how to reproduce it on demand, otherwise we could bisect 1.21 -> 1.22 to try to figure out the commit.

**Comment by Stono** [comment]
Hey, no @jpuskar i didn't get anywhere, and it hasn't come back recently either (and we're on 1.23 now).
Really bothers me that I can't explain it.

**Comment by Stono** [comment]
Hey, 
So we never got to the bottom of it, however we had similar issues (such as https://github.com/istio/istio/issues/33466#issuecomment-1100164820).

We had a somewhat working theory / best guess that perhaps this was due to dead connections being kept open by envoy/the kernel.

We now set `sysctl -w net.ipv4.tcp_retries2=5` - and i haven't seen it recently.  However:

- It was extremely sporadic anyway
- We're on istio `1.24.1` now
- We've done numerous kubernetes upgrades since raising this
- We've released many other software changes to the apps in this time

So i've no clue if it's what helped, or not.
